File: 2023/go/cmd/main.go
package main

import (
	"flag"
	"fmt"
	"os"
	"strings"

	"advent/pkg/day1"
	"advent/pkg/day2"
	"advent/pkg/day3"
	"advent/pkg/day4"
	"advent/pkg/day5"
	"advent/pkg/day6"
	"advent/pkg/day7"
	"advent/pkg/day8"
)

func main() {
	dayPtr := flag.Int("day", 1, "day to run")
	problemPtr := flag.Int("problem", 1, "problem to run")
	flag.Parse()

	file, err := os.ReadFile(fmt.Sprintf("/home/jan/advent/2023/go/in/day%d.in", *dayPtr))

	if err != nil {
		panic(err)
	}

	lines := strings.Split(string(file), "\n")

	result := "Problem doesn't exist"

	fmt.Printf("Solving day %d, problem %d\n", *dayPtr, *problemPtr)
	switch *dayPtr {
	case 1:
		switch *problemPtr {
		case 1:
			result = day1.SolveProblem1(lines)
		case 2:
			result = day1.SolveProblem2(lines)
		}

	case 2:
		switch *problemPtr {
		case 1:
			result = day2.SolveProblem1(lines)
		case 2:
			result = day2.SolveProblem2(lines)
		}

	case 3:
		switch *problemPtr {
		case 1:
			result = day3.SolveProblem1(lines)
		case 2:
			result = day3.SolveProblem2(lines)
		}

	case 4:
		switch *problemPtr {
		case 1:
			result = day4.SolveProblem1(lines)
		case 2:
			result = day4.SolveProblem2(lines)
		}

	case 5:
		switch *problemPtr {
		case 1:
			result = day5.SolveProblem1(lines)
		case 2:
			result = day5.SolveProblem2(lines)
		}

	case 6:
		switch *problemPtr {
		case 1:
			result = day6.SolveProblem1(lines)
		case 2:
			result = day6.SolveProblem2(lines)
		}

	case 7:
		switch *problemPtr {
		case 1:
			result = day7.SolveProblem1(lines)
		case 2:
			result = day7.SolveProblem2(lines)
		}

	case 8:
		switch *problemPtr {
		case 1:
			result = day8.SolveProblem1(lines)
		case 2:
			result = day8.SolveProblem2(lines)
		}
	default:
		result = "Invalid day number"
	}

	fmt.Println(result)
}

File: 2023/rust/src/main.rs
use std::fs;

mod day3;

fn main() {
    let args: Vec<String> = std::env::args().collect();

    if args.len() < 2 {
        println!("Please provide a day number and the problem number as arguments");
        return;
    }

    let day = args[1]
        .parse::<i32>()
        .expect("Failed to parse day number :(");
    let problem = args[2]
        .parse::<i32>()
        .expect("Failed to parse problem number :(");

    let file = fs::read_to_string(format!("in/day{day}.in")).expect("Failed to read file :(");
    let lines: Vec<&str> = file.lines().collect();

    println!("Day {}, problem {}", day, problem);

    let result = match day {
        3 => match problem {
            1 => day3::problem1::solve(lines),
            2 => day3::problem2::solve(lines),
            _ => panic!("Invalid problem number :("),
        },
        _ => panic!("Invalid day number :("),
    };

    println!("{}", result);
}
